### DISCUSSION TITLE
Howto provide compiler flags in cmake

### DISCUSSION THREAD

**Discussion by aronroland** [NOAA-EMC/WW3#822] [Q&A] [answered]
Dear All, 

do we have a unified way how we prescribe compiler flags in cmake? Can anybody explain how to provide compiler flags in cmake? Do we have default compiler flags for debug & release for each compiler? 

Cheers

Aron

**Comment by mickaelaccensi** [comment, marked as answer]
Hi Aron,

You can specify the following option to compile with debug options : -DCMAKE_BUILD_TYPE=Debug
The documentation is still not released, however I've done a tutorial about compilation witch cmake for the training course. 
It is available here : https://docs.google.com/document/d/1azvnWsozWDpSo6c7gSFwZVAsqQ5wcpLndsXDXe1W4gY/edit?usp=sharing

Let us know if you need more help,

Cheers,

Mickael

**Reply by aronroland** [reply]
Hi Mickael, 

thanks for the answer and sharing your doc. The doc is fine i just looked at it but I am still missing the way howto prescribe the compiler flags. I know how to choose the debug or release version but we must also be able to set the flags. It would be really nice if you could extend your tutorial in this direction. 

Cheers

Aron

**Reply by aronroland** [reply]
To change compiler flags, "ccmake ." can be used in the build directory. Please check ccmake documentation...

**Reply by ukmo-ccbunney** [reply]
Thanks for pointing that out @aronroland - that's a useful tool!

**Reply by aronroland** [reply]
so yesterday it was a bit late ... so give here some super short manual ... for those who r lazy like i am ... 
(t) to get an extended list of options
(enter) to change the value of a certain parameter
(c) to save the configuration 
(q) to exit ccmake ... 

normal i do: (t) ... i do all my changes then (c) and exit (q) thats it ... but now it becomes a bit nasty since I will now create a new issue about compiler flags ...

**Comment by ukmo-ccbunney** [comment]
Hey all 
Just found something out that might be of interest with respect to setting compiler flags.
You can pass in a "toolchain" file to cmake that you can use to set specific cmake variables for a build. With this, you can override the  fortran flags for a build config. I've successfully got this to work for a build using the nvidia compiler. 

Example:

**file: NVToolchain.make**
```cmake
# override DEBUG and RELEASE Fortran flags:
set(CMAKE_Fortran_FLAGS_DEBUG "-g -O0 -Mbounds -acc -Minfo=acc -DW3_GPU -ta=tesla,managed" CACHE STRING "")
set(CMAKE_Fortran_FLAGS_RELEASE "-O2 -acc -Minfo=acc -DW3_GPU -ta=tesla,managed" CACHE STRING "")
```

Now call cmake as usual, but with the `-DCMAKE_TOOLCHAIN_FILE` option:

```
cmake .. -DSWITCH=${SWITCH_FILE} \
        -DCMAKE_TOOLCHAIN_FILE:string=NVToolchain.cmake \    # <--- specify toolchain
        -DCMAKE_INSTALL_PREFIX=install
```

Running `ccmake` after this should confirm that your flags are set to what is in the toolchain file.

I guess this is useful as you can have seperate toolchain files for different compiler/system combinations?

I thought it would be worth mentioning....

**Reply by MatthewMasarik-NOAA** [reply]
This is very cool to be able to override settings in the `CMakeLists.txt`.  I've been playing with some alternate compiler options (from @aronroland) for the SCOTCH work, and had been making edits to the cmake standard and debug compile flag portions.  This toolchain file method avoids making direct edits to `CMakeLists.txt`, which is much better imo, given the potential for issues if the edit's to `CMakeLists.txt` are not properly tracked.   @ukmo-ccbunney , thanks for sharing!
